pallets/phala-world/src/lib.rs

```diff
@@ -439,133 +510,317 @@ pub mod pallet {
+			let origin_of_shell_collection_id = OriginOfShellCollectionId::<T>::get()
+				.ok_or(Error::<T>::OriginOfShellCollectionNotSet)?;
+			// Must have a spirit to purchase a Origin of Shell
+			ensure!(
+				pallet_uniques::Pallet::<T>::owned_in_class(&spirit_collection_id, &sender).count() >
+					0,
+				Error::<T>::MustOwnSpiritToPurchase
+			);
+			// If not last day of sale, only allowed to purchase one Origin of Shell
+			ensure!(
+				(!LastDayOfSale::<T>::get() &&
+					pallet_uniques::Pallet::<T>::owned_in_class(
+						&origin_of_shell_collection_id,
+						&sender
+					)
+					.count() == 0) || LastDayOfSale::<T>::get(),
+				Error::<T>::OriginOfShellAlreadyPurchased
+			);
+			// Get Origin of Shell Price based on Origin of ShellType
+			let origin_of_shell_price = match origin_of_shell_type {
+				OriginOfShellType::Legendary => T::LegendaryOriginOfShellPrice::get(),
+				OriginOfShellType::Magic => T::MagicOriginOfShellPrice::get(),
 				_ => return Err(Error::<T>::InvalidPurchase.into()),
 			};
-			let nft_id = pallet_rmrk_core::NextNftId::<T>::get(egg_collection_id);
+			let nft_id = pallet_rmrk_core::NextNftId::<T>::get(origin_of_shell_collection_id);
 			// Check if race and career types have mints left
-			Self::has_race_type_left(&race)?;
-			Self::has_career_type_left(&career)?;
+			Self::has_race_type_left(&origin_of_shell_type, &race)?;
 
-			// Define EggInfo for storage
-			let egg = EggInfo {
-				egg_type,
+			// TODO: Update this for incubation info Define OriginOfShellInfo for storage
+			let origin_of_shell = OriginOfShellInfo {
+				origin_of_shell_type: origin_of_shell_type.clone(),
 				race: race.clone(),
 				career: career.clone(),
-				start_hatching: 0,
-				hatching_duration: 0,
+				start_incubation: 0,
+				incubation_duration: 0,
 			};
 
-			// Transfer the amount for the rare Egg NFT then mint the egg
+			// Transfer the amount for the rare Origin of Shell NFT then mint the origin_of_shell
 			<T as pallet::Config>::Currency::transfer(
 				&sender,
 				&overlord,
-				egg_price,
+				origin_of_shell_price,
 				ExistenceRequirement::KeepAlive,
 			)?;
-			// Mint Egg and transfer Egg to new owner
+			// Mint Origin of Shell and transfer Origin of Shell to new owner
 			pallet_rmrk_core::Pallet::<T>::mint_nft(
-				Origin::<T>::Signed(overlord.clone()).into(),
+				Origin::<T>::Signed(overlord).into(),
 				sender.clone(),
-				egg_collection_id,
+				origin_of_shell_collection_id,
 				None,
 				None,
 				metadata,
 			)?;
+			// Set Origin of Shell Type, Race and Career attributes for NFT
+			Self::set_race_and_career_attributes(
+				origin_of_shell_collection_id,
+				nft_id,
+				race.clone(),
+				career.clone(),
+			)?;
 
-			Self::decrement_race_type(race);
-			Self::decrement_career_type(career);
-			Eggs::<T>::insert(egg_collection_id, nft_id, egg);
+			// Update storage
+			Self::decrement_race_type_left(origin_of_shell_type.clone(), race.clone())?;
```

Comment:
Is the "left" and "race type (purchased)" redundant? If we know the total and purchase, we can easily derive the remaining amount

Reply:
I had created a struct to hold `NftSaleInfo`
```
/// NftSaleInfo is used as the value in the StorageDoubleMap that takes key1 as the
/// OriginOfShellType and key2 as the RaceType
#[derive(Encode, Decode, Eq, PartialEq, Clone, RuntimeDebug, TypeInfo, MaxEncodedLen)]
pub struct NftSaleInfo {
	/// Number of Race Type count
	pub race_count: u32,
	/// Number of races left to sell
	pub race_for_sale_count: u32,
	/// Number of giveaway races left
	pub race_giveaway_count: u32,
	/// Number of reserved races left
	pub race_reserved_count: u32,
}
```
If this is too much, then I can track just the number purchased and minted.